// funclib/lib/index.js
module.exports = {
  typeOf: require('./typeOf'),
  typeVal: require('./typeVal'),
  isStr: require('./isStr'),
  isNum: require('./isNum'),
  isBol: require('./isBol'),
  isFun: require('./isFun'),
  isNul: require('./isNul'),
  isUdf: require('./isUdf'),
  isErr: require('./isErr'),
  isDat: require('./isDat'),
  isReg: require('./isReg'),
  isArr: require('./isArr'),
  isObj: require('./isObj'),
  array: require('./array'),
  range: require('./range'),
  toArr: require('./toArr'),
  indexOf: require('./indexOf'),
  find: require('./find'),
  filter: require('./filter'),
  reject: require('./reject'),
  contains: require('./contains'),
  drop: require('./drop'),
  flatten: require('./flatten'),
  pluck: require('./pluck'),
  uniq: require('./uniq'),
  forEach: require('./forEach'),
  each: require('./forEach'),
  sortBy: require('./sortBy'),
  len: require('./len'),
  has: require('./has'),
  get: require('./get'),
  set: require('./set'),
  keys: require('./keys'),
  pick: require('./pick'),
  omit: require('./omit'),
  extend: require('./extend'),
  forIn: require('./forIn'),
  deepCopy: require('./deepCopy'),
  isEmpty: require('./isEmpty'),
  isDeepEqual: require('./isDeepEqual'),
  random: require('./random'),
  randomId: require('./randomId'),
  randomColor: require('./randomColor'),
  interval: require('./interval'),
  timeout: require('./timeout'),
  defer: require('./defer'),
  time: require('./timestamp'),
  timestamp: require('./timestamp'),
  asUtcTime: require('./asUtcTime'),
  asXyzTime: require('./asXyzTime'),
  fmtDate: require('./fmtDate'),
  fmtUtcDate: require('./fmtUtcDate'),
  fmtXyzDate: require('./fmtXyzDate'),
  match: require('./match'),
  pretty: require('./pretty'),
  escape: require('./escape'),
  unescape: require('./unescape'),
  capitalize: require('./capitalize'),
  fmtCurrency: require('./fmtCurrency'),
  maskString: require('./maskString'),
  cutString: require('./cutString'),
  parseQueryStr: require('./parseQueryStr'),
  stringifyQueryStr: require('./stringifyQueryStr'),
  setPattern: require('./setPattern'),
  getPattern: require('./getPattern'),
  testPattern: require('./testPattern'),
  matchPattern: require('./matchPattern'),
  rest: require('./rest'),
  throttle: require('./throttle'),
  debounce: require('./debounce'),
  chalk: require('./chalk'),
  print: require('./printS'),
  log: require('./logS'),
  rd: require('./rd'),
  wt: require('./wt'),
  cp: require('./cp'),
  mv: require('./mv'),
  rm: require('./rm'),
  mk: require('./mk'),
  size: require('./size'),
  clear: require('./clear'),
  progress: require('./progress'),
  version: require('./_config').version,
}
